Validate the allocation before adding it to the registry

WilliamMcCumstie opened the pull request:
RONSEAL

WilliamMcCumstie committed:
Log when an persisted allocation is missing its job

This shouldn't happen often, however it may occur when depending
when a restart occurs

WilliamMcCumstie committed:
Prevent invalid allocations being added to the registry

WilliamMcCumstie committed:
Add validations on to the Allocation model

benarmston commented on app/models/allocation.rb:
```diff
@@ -35,14 +35,28 @@ class  MissingNodeError < RuntimeError; end
   end
 
+  validate do
+    if nodes.is_a? Array && nodes.any?
+      errors.add :nodes, 'Must only contain nodes' unless nodes.all? { |n| n.is_a? Node }
```
As a general rule I don't like using `any?` as a substitute for `! empty?`.  They don't behave the same and it always leaves me wondering if the difference is important.

```
irb(main):028:0* [].any? == ! [].empty?
=> true
irb(main):029:0> [nil].any? == ! [nil].empty?
=> false
```

---

We could combine the `any?` (`! empty?`) check with the `all?` check to have one less conditional to read.

```
if nodes.is_a? Array && ! nodes.all? { |n| n.is_a? Node }
```

WilliamMcCumstie replied:
This conditional has changed quite a bit since your review:
https://github.com/openflighthpc/flight-scheduler-controller/pull/48/commits/85214b58f53741d7297b034b3e2a46bc5bedd7b4

benarmston commented on app/models/allocation.rb:
```diff
@@ -51,6 +65,10 @@ def initialize(job:, nodes:)
     end
   end
 
+  def valid?
+    true
+  end
```
Why do we have this?

WilliamMcCumstie replied:
https://github.com/openflighthpc/flight-scheduler-controller/pull/48/commits/85214b58f53741d7297b034b3e2a46bc5bedd7b4

benarmston commented on lib/flight_scheduler/allocation_registry.rb:
```diff
@@ -201,13 +208,21 @@ def load
-          end
-          @job_allocations[allocation.job.id] = allocation
+      allocations = data.map { |h| [h['job_id'], Allocation.from_serialized_hash(h)] }
+      bad, good = allocations.partition { |_, a| a.job.nil? }
```
As you've added validation, should we not be partitioning on the allocation being valid rather than just the case of the job being nil?

Alternatively, we could blindly add the allocations to the registry and catch the `AllocationConflict`.  Perhaps its better to ask for forgiveness here rather than permission.

WilliamMcCumstie replied:
https://github.com/openflighthpc/flight-scheduler-controller/pull/48/commits/a3923a7edde5b6eb97427e23705cba7819f65344

benarmston reviewed (changes requested):
Looks like you've left in a bit of WIP.  I've also added a couple of comments that I'd like you to look at.

WilliamMcCumstie committed:
Reduce the validation logic in Allocation

WilliamMcCumstie committed:
Ensure allocations are valid when added to registry

WilliamMcCumstie committed:
Use a factory to build Allocations

This is to allow the allocation model to be refactored easily

WilliamMcCumstie committed:
Initialize allocations with node_names not nodes

This removes the logic from the constructor

WilliamMcCumstie committed:
Create missing nodes within AllocationRegistry

This only needs to occur because the allocation registry's
persistence may contain references to nodes which no longer
exist

This is the only time missing nodes should be created

WilliamMcCumstie committed:
Used the validation method when loading allocations

benarmston commented on app/models/allocation.rb:
```diff
@@ -35,35 +35,39 @@ class  MissingNodeError < RuntimeError; end
+  end
+
+  # TODO: Remove add flag
   def nodes(add: false)
```
What's stopping this from being done now?

WilliamMcCumstie replied:
Hold over from refactoring, all the calls have previously been removed:
https://github.com/openflighthpc/flight-scheduler-controller/pull/48/commits/f240454c965a10f05bc24eb30fa789912b778721

benarmston reviewed (changes requested):
It looks like you've left a `XXX` which can be done now.  Could you take a look.

WilliamMcCumstie committed:
Remove the old add flag entirely

benarmston commented:
LGTM.
